### ISSUE TITLE
Accommodate app / asset / buildspec / appspec files outside of cdk folder in pipeline?

### ISSUE THREAD

**Issue by jewelsjacobs** [eladb/cdkcd-test#1]
First off this is amazing. I can't think of a better way to accommodate a multistack pipeline.

My only wish list would be the ability to access files outside of the cdk folder, which in this example is the root of this cdkcd-test repo.

What do I mean?  Here is the use case (which I'm dealing with right now):

I have a front end app that gets build with webpack and I was asked to have the cdk folder in the same directory (top lvl). It looks like this:

```
├── dist <- compiled static files from build
├── src <- files webpack builds
└── infrastructure <- cdk folder with cloudfront / route53 service stack and pipeline stack with codebuild / codepipeline that also creates an s3 bucket. 
```

The pipeline stack includes an over complicated buildspec that:
-  builds the static app files
-  runs an aws s3 sync cli command
-  synthesizes the cloudfront / route53 service stack.

The process of having to build static files before deploying infrastructure stacks is clunky at best and I'm not sure the best way to handle it TBH.

I know I'm supposed to use a shell script in a buildspec to accommodate relative folder structures but if this DeployAction includes a project which includes the buildspec can abstract something like this CDK would be even more awesome!

TL;DR; Where do I request a "build static assets and set as S3DeployAction input" in this fabulous pipeline without pain?

Does this make any sense?

**Comment by eladb** [comment]
Hey, thanks for the feedback here! I am wondering if you have looked into using the @aws-cdk/aws-s3-deployment module to upload you assets to s3? I think this would have streamlined your pipeline....

**Comment by jewelsjacobs** [comment]
Thanks for responding to the issue! I didn't use the @aws-cdk/aws-s3-deployment module (which I use in other cases) because I thought I couldn't reference asset paths in the parent folder but after doing a test I see that you can!

Awesome!

This will allow me to create a seperate S3 project source to assign a buildspec to :)

I'll try this!

**Comment by jewelsjacobs** [comment]
The problem though is this:

I still have to *build / compile* files to the dist dir *BEFORE* *sync*'ing it to the s3 bucket. Little bit of a chicken before the egg issue. 

That requires putting compile commands somewhere before using the @aws-cdk/aws-s3-deployment module to sync.

My workaround is to put something like this in the project buildspec:

```
          build: {
            commands: [
                'echo Entered the build phase',
                'yarn run build', <- build static files from src to dist
                'echo "Uploading build to $STATIC_SITE_BUCKET ..."',
                'aws s3 sync "./dist" "$STATIC_SITE_BUCKET" --delete', <- sync files
                'cd infrastructure',
                'yarn run build', <- compile cdk files (using yarn with cdk here)
                `yarn run cdk synth ${props.serviceStack} -o ./dist`, <- synth the cdk template
                'cd ..',
                `cp -r infrastructure/dist/*.template.json ./dist` <put everything in the same artifact base directory
            ],
          }
```

I can't sync the `src` directory with  @aws-cdk/aws-s3-deployment module because I still need other files to compile those files. 

I can't figure out a more elegant way to do this. Any ideas?

**Comment by jewelsjacobs** [comment]
FYI: the current solution I am using is is webpack specific. I'm using the [webpack-s3-plugin
](https://github.com/MikaAK/s3-plugin-webpack).

Adding it into the webpack config:

```
new S3Plugin({
    include: /.*\.(css|js|eot|otf|ttf|woff|woff2|jpe?g|svg|png|gif|map)/,
    s3Options: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    },
    s3UploadOptions: {
      Bucket: process.env.S3_BUCKET,
    },
    progress: false,
  });
```

That brings up the challenge of having the pass those aws credentials through the buildpack.

I see you are doing that in your [deploy-action.ts](https://github.com/eladb/cdkcd-test/blob/master/lib/deploy-action.ts#L59-L64)

It seems, though, this requires the creation of a **third party aws-cdk-deployment role** to create temp credentials. I'm also not sure if that `const EXTERNAL_ID = 'AWS::CDK::DEPLOY::021D8AD7C55D49598C82ACB6BFC30F1B'` can be reused.

I wonder if there is anyway CDK can create this so it doesn't have to be created manually. Ill add that question / info to the RFC.
